_posts/tutorials/2021-04-16-javalin-sureness-example.md

```diff
@@ -0,0 +1,298 @@
+author: <a href="https://github.com/tomsun28" target="_blank">Tom</a>
+date: 2021-04-16
+summarytitle: Using Sureness to protect the security of Javalin REST API.
+summary: Using Sureness to secure Javalin REST API by providing authentication(jwt,bsaic,digest) and authorization(rbac)"
+language: Java
+---
+
+## What is Sureness   
+
+> [Sureness](https://github.com/usthe/sureness) is a simple and efficient open-source security framework that focus on protection of restful api.  
+> Provide authentication and authorization, based on RBAC.   
+> No specific framework dependency(support javalin, springboot, quarkus, ktor and more).    
+> Support dynamic modification of permissions.   
+> Support websocket, mainstream http container(servlet and jax-rs).    
+> Supports JWT, Basic Auth, Digest Auth... Can extend custom supported authentication methods.    
+> High performance due dictionary matching tree.      
+> Good extension interface, demo and document.    
+
+> The low configuration of sureness, easy to expand, and not coupled with other frameworks, enables developers to quickly and safely protect their projects in multiple scenarios.   
+
+
+## What You Will Learn
+
+* Creating a simple REST API using Javalin  
+* Learn how to integrate Sureness into a Javalin application  
+* Learn how to issue a jwt  
+* Test API authentication - use jwt auth, basic auth, digest auth to test the security of the REST API   
+* Test API authorization - use different users to verify that they can access the REST API   
+
+
+The tutorial assumes that you know what jwt, basic auth, digest auth, rbac are. If you 
+do not, then you can check [jwt](https://jwt.io/introduction/), [basic auth](https://docs.oracle.com/cd/E50612_01/doc.11122/user_guide/content/authn_http_basic.html) , [digest auth](https://docs.oracle.com/cd/E50612_01/doc.11122/user_guide/content/authn_http_digest.html), [rbac](https://en.wikipedia.org/wiki/Role-based_access_control) for an introduction.
+
+## Setting Up Dependencies
+
+First, you will need to create a maven project and add javalin, sureness dependencies coordinate   
+
+````xml
+        <dependency>
+            <groupId>io.javalin</groupId>
+            <artifactId>javalin</artifactId>
+            <version>3.10.1</version>
+        </dependency>
+        <dependency>
+            <groupId>com.usthe.sureness</groupId>
+            <artifactId>sureness-core</artifactId>
+            <version>1.0.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
+            <version>1.7.30</version>
+        </dependency>
+````
+
+
+## Setting Up Javalin and create REST API   
+
+We need to create a simple Javalin app and provide some  REST API for test.    
+
+```
+        // init javalin
+        Javalin app = Javalin.create().start(8088);
+```
+
+```java
+        // create simple rest api
```

Comment:
This is okay, but a more idiomatic way of writing this in Javalin would be:

```java
app.routes(() -> {
    path("api", () -> {
        path("v3", () -> {
            get("host", ctx -> ctx.result("get /api/v3/host success"));
        });
        path("v2", () -> {
            path("host", () -> {
                get(ctx -> ctx.result("get /api/v2/host success"));
                post(ctx -> ctx.result("post /api/v2/host success"));
                put(ctx -> ctx.result("put /api/v2/host success"));
                delete(ctx -> ctx.result("delete /api/v2/host success"));
            });
            get("getSource2/book", ctx -> ctx.result("get /api/v2/getSource2/book success"));
        });
        path("v1", () -> {
            get("getSource1", ctx -> ctx.result("get /api/v1/getSource1 success"));
            path("source1", () -> {
                get(ctx -> ctx.result("get /api/v1/source1 success"));
                post(ctx -> ctx.result("post /api/v1/source1 success"));
                put(ctx -> ctx.result("put /api/v1/source1 success"));
                delete(ctx -> ctx.result("delete /api/v1/source1 success"));
            });
        });
        path("mi", () -> {
            put("tom", ctx -> ctx.result("put /api/mi/tom success"));
        });
    });
```

This way all the String repetition is eliminated, and it's really easy to see which controllers belong to which versions of the API.